d2layouts/d2elklayout/layout.go

```diff
@@ -442,5 +443,270 @@ func Layout(ctx context.Context, g *d2graph.Graph, opts *ConfigurableOpts) (err
 		edge.Route = points
 	}
 
+	deleteBends(g)
+
 	return nil
 }
+
+// deleteBends is a shim for ELK to delete unnecessary bends
+// see https://github.com/terrastruct/d2/issues/1030
+func deleteBends(g *d2graph.Graph) {
+	// Get rid of S-shapes at the source and the target
+	// TODO there might be value in repeating this. removal of an S shape introducing another S shape that can still be removed
+	for _, isSource := range []bool{true, false} {
+		for ei, e := range g.Edges {
+			if len(e.Route) < 4 {
+				continue
+			}
+			if e.Src == e.Dst {
+				continue
+			}
+			var endpoint *d2graph.Object
+			var start *geo.Point
+			var corner *geo.Point
+			var end *geo.Point
+
+			if isSource {
+				start = e.Route[0]
+				corner = e.Route[1]
+				end = e.Route[2]
+				endpoint = e.Src
+			} else {
+				start = e.Route[len(e.Route)-1]
+				corner = e.Route[len(e.Route)-2]
+				end = e.Route[len(e.Route)-3]
+				endpoint = e.Dst
+			}
+
+			isHorizontal := math.Ceil(start.Y) == math.Ceil(corner.Y)
+
+			// Make sure it's still attached
+			if isHorizontal {
+				if end.Y <= endpoint.TopLeft.Y+10 {
+					continue
+				}
+				if end.Y >= endpoint.TopLeft.Y+endpoint.Height-10 {
+					continue
+				}
+			} else {
+				if end.X <= endpoint.TopLeft.X+10 {
+					continue
+				}
+				if end.X >= endpoint.TopLeft.X+endpoint.Width-10 {
+					continue
+				}
+			}
+
+			var newStart *geo.Point
+			if isHorizontal {
+				newStart = geo.NewPoint(start.X, end.Y)
+			} else {
+				newStart = geo.NewPoint(end.X, start.Y)
+			}
+
+			endpointShape := shape.NewShape(d2target.DSL_SHAPE_TO_SHAPE_TYPE[strings.ToLower(endpoint.Attributes.Shape.Value)], endpoint.Box)
+			newStart = shape.TraceToShapeBorder(endpointShape, newStart, end)
+
+			// Check that the new segment doesn't collide with anything new
+
+			oldSegment := geo.NewSegment(start, corner)
+			newSegment := geo.NewSegment(newStart, end)
+
+			oldIntersects := countObjectIntersects(g, e.Src, e.Dst, *oldSegment)
+			newIntersects := countObjectIntersects(g, e.Src, e.Dst, *newSegment)
+
+			if newIntersects > oldIntersects {
+				continue
+			}
+
+			oldCrossingsCount, oldOverlapsCount, oldCloseOverlapsCount, oldTouchingCount := countEdgeIntersects(g, g.Edges[ei], *oldSegment)
+			newCrossingsCount, newOverlapsCount, newCloseOverlapsCount, newTouchingCount := countEdgeIntersects(g, g.Edges[ei], *newSegment)
+
+			if newCrossingsCount > oldCrossingsCount {
+				continue
+			}
+			if newOverlapsCount > oldOverlapsCount {
+				continue
+			}
+
+			if newCloseOverlapsCount > oldCloseOverlapsCount {
+				continue
+			}
+			if newTouchingCount > oldTouchingCount {
+				continue
+			}
```

Comment:
it seems like we only need the overlaps count since it covers close overlaps and touching

Reply:
we don't want it to get progressively worse